packages/react-router/modules/Router.js

```diff
@@ -16,17 +16,34 @@ class Router extends React.Component {
 
   getChildContext() {
     return {
-      router: this.props.history
+      router: this.router
     }
   }
 
   componentWillMount() {
-    const { children } = this.props
+    const { children, history } = this.props
 
     invariant(
       children == null || React.Children.count(children) === 1,
       'A <Router> may have only one child element'
     )
+
+    this.router = {
+      ...history,
+      match: {
+        path: '/',
+        url: '/',
+        params: {},
+        isExact: history.location.pathname === '/'
+      }
+    }
+
+    history.listen(() => {
+      Object.assign(this.router, history)
+      Object.assign(this.router.match, {
```

Comment:
If you tried to combine these `Object.assign`s, it would end up overwriting the deep `match` property, rather than merging it.